File: main/binary-tree-postorder-traversal/binary-tree-postorder-traversal-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(const TreeNode* root);
};

vector<int> Solution::postorderTraversal(const TreeNode* root)
{
    vector<int> ret;
    
    const TreeNode* last {};
    stack<const TreeNode*> s;
    
    while (root || !s.empty()) {
        // Go left as far as possible.
        for (; root; root = root->left) s.push(root);
        
        // Go right from here by one, if we can and haven't already.
        if (const auto right = s.top()->right) {
            if (right != last) {
                root = right;
                continue;
            }
        }
        
        // Use the value here and retreat.
        last = s.top();
        s.pop();
        ret.push_back(last->val);
    }
    
    return ret;
}

File: main/binary-tree-postorder-traversal/binary-tree-postorder-traversal-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        final List<Integer> vals = new ArrayList<>();
        final Deque<TreeNode> stack = new ArrayDeque<>();
        TreeNode post = null; // The last node retreated from, if any.

        while (root != null || !stack.isEmpty()) {
            // Go left as far as possible.
            for (; root != null; root = root.left) stack.push(root);

            final TreeNode cur = stack.element();

            if (cur.right != null && cur.right != post) {
                // The right subtree is nonempty and unvisited. Go there.
                root = cur.right;
            } else {
                // The right subtree is explored. Display and retreat.
                post = cur;
                vals.add(post.val);
                stack.pop();
            }
        }

        return vals;
    }
}
